File: 0905. Sort Array By Parity/0905.java
// time complexity: O(n)
// space complexity: O(1)
class Solution {
    // two-pointer in-place swap
    public int[] sortArrayByParity(int[] nums) {
        int n = nums.length;
        int j = n - 1;                   // initialize right pointer
        for (int i = 0; i < n; i++) {
            if (nums[i] % 2 == 1) {      // if current is odd
                if (nums[j] % 2 == 0) {      // find an even at right
                    int temp = nums[i];
                    nums[i] = nums[j];       // swap odd-left with even-right
                    nums[j] = temp;
                } else {
                    i--;                   // stay on this i to retry
                }
                j--;                       // move right pointer inward
            }
            if (i == j) break;            // stop when pointers meet
        }
        return nums;                      // return partitioned array
    }
}

File: 0905. Sort Array By Parity/0905.py
# time complexity: O(n)
# space complexity: O(n)
class Solution(object):
    # list comprehension, new list
    def sortArrayByParity(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        # build list: evens then odds
        evens = [x for x in nums if x % 2 == 0]
        odds  = [x for x in nums if x % 2 != 0]
        return evens + odds